### PR TITLE
btrfs: Do not disable quota on cleanup

### LINKED ISSUE #78: Docker disables btrfs quotas creating an unrecoverable broken btrfs volume
* [x] This is a bug report
* [ ] This is a feature request
* [x] I searched existing issues before opening this one

### Expected behavior
When using docker on a btrfs volume where the quotas are enabled and certain limits on certain qgroups are set those should be retained and still be there after a reboot of the system.

### Actual behavior
When using docker on a btrfs volume where quotas are enabled and certain limits on certain qgroups are set the btrfs volume gets corrupted in a way that seems unrecoverable as soon as you start the first container. This behavior does not show immediately but after a reboot the quotas are disabled and the qgroup limits are gone. I found no other way to recover from that state except to reformat the btrfs volume.

### Steps to reproduce the behavior

**Output of `docker version`:**

```
Client:
 Version:      17.06.0-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   3dfb8343
 Built:        Wed Jul 26 18:03:33 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.06.0-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   02c1d87617
 Built:        Wed Jul 26 20:03:39 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 1
 Running: 0
 Paused: 0
 Stopped: 1
Images: 1
Server Version: 17.06.0-ce
Storage Driver: btrfs
 Build Version: Btrfs v4.11.1
 Library Version: 102
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: cfb82a876ecc11b5ca0977d1733adbe58599088a
runc version: 2d41c047c83e09a6d61d464906feb2a2f3c52aa4
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.12.4-1-ARCH
Operating System: Arch Linux
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 7.686GiB
Name: peregrin
ID: WGK4:K2IQ:GTX5:JMQQ:45EB:T3MU:OA62:TIUD:VIMX:VV3P:OISY:IRC6
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

Steps to reproduce (run all commands as root for simplicity):
1. Install Linux using btfs root (I use arch linux)
2. enable btrfs quotas
```
btrfs quota enable /
```
3. Create a btrfs subvolume
```
btrfs subvolume create /sub1
```
4. check quota
```
btrfs qgroup show -f -re /sub1 
qgroupid         rfer         excl     max_rfer     max_excl 
--------         ----         ----     --------     -------- 
0/265        16.00KiB     16.00KiB         none         none
```

5. set a limit for the new subvolume
```
btrfs qgroup limit 10M 0/265 /
```
6. check quota
```
btrfs qgroup show -f -re /sub1 
qgroupid         rfer         excl     max_rfer     max_excl 
--------         ----         ----     --------     -------- 
0/265        16.00KiB     16.00KiB     10.00MiB         none 
```
=> You see the limit is set!

7. install docker
8. run any docker container (e.g busybox)
```
docker run -ti busybox
```
9. check quota
```
btrfs qgroup show -f -re /sub1 
qgroupid         rfer         excl     max_rfer     max_excl 
--------         ----         ----     --------     -------- 
0/265        16.00KiB     16.00KiB     10.00MiB         none 
```
=> All seems still fine up to here.
10. reboot the system
11. check quota
```
btrfs qgroup show -f -re /sub1
ERROR: can't list qgroups: quotas not enabled
```
=> Quota support is gone.
12. re-enable quota to see what it does
```
btrfs quota enable /
```
13. check quotas
```
btrfs qgroup show -f -re /sub1
qgroupid         rfer         excl     max_rfer     max_excl 
--------         ----         ----     --------     -------- 
0/265        16.00KiB     16.00KiB         none         none
```
=> The previously set quota is gone!

### LINKED ISSUE #34593: Docker's btrfs graph driver eliminates the existing quota controls
**Description**

**Steps to reproduce the issue:**
1. Install Linux using btfs root (I use arch linux)
2. enable btrfs quotas
```
btrfs quota enable /
```
3. Create a btrfs subvolume
```
btrfs subvolume create /sub1
```
4. check quota
```
btrfs qgroup show -f -re /sub1 
qgroupid         rfer         excl     max_rfer     max_excl 
--------         ----         ----     --------     -------- 
0/265        16.00KiB     16.00KiB         none         none
```
5. set a limit for the new subvolume
```
btrfs qgroup limit 10M 0/265 /
```
6. check quota
```
btrfs qgroup show -f -re /sub1 
qgroupid         rfer         excl     max_rfer     max_excl 
--------         ----         ----     --------     -------- 
0/265        16.00KiB     16.00KiB     10.00MiB         none 
```
=> You see the limit is set!

7. install docker
8. run any docker container (e.g busybox)
```
docker run -ti busybox
```
9. check quota
```
btrfs qgroup show -f -re /sub1 
qgroupid         rfer         excl     max_rfer     max_excl 
--------         ----         ----     --------     -------- 
0/265        16.00KiB     16.00KiB     10.00MiB         none 
```
=> All seems still fine up to here.
10. reboot the system
11. check quota
```
btrfs qgroup show -f -re /sub1
ERROR: can't list qgroups: quotas not enabled
```
=> Quota support is gone.
12. re-enable quota to see what it does
```
btrfs quota enable /
```
13. check quotas
```
btrfs qgroup show -f -re /sub1
qgroupid         rfer         excl     max_rfer     max_excl 
--------         ----         ----     --------     -------- 
0/265        16.00KiB     16.00KiB         none         none
```
=> The previously set quota is gone!

**Describe the results you received:**
When using docker on a btrfs volume where quotas are enabled and certain limits on certain qgroups are set the btrfs volume gets corrupted in a way that seems unrecoverable as soon as you start the first container. This behavior does not show immediately but after a reboot the quotas are disabled and the qgroup limits are gone. I found no other way to recover from that state except to reformat the btrfs volume.

**Describe the results you expected:**
When using docker on a btrfs volume where the quotas are enabled and certain limits on certain qgroups are set those should be retained and still be there after a reboot of the system.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      17.06.0-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   3dfb8343
 Built:        Wed Jul 26 18:03:33 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.06.0-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   02c1d87617
 Built:        Wed Jul 26 20:03:39 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 1
 Running: 0
 Paused: 0
 Stopped: 1
Images: 1
Server Version: 17.06.0-ce
Storage Driver: btrfs
 Build Version: Btrfs v4.11.1
 Library Version: 102
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: cfb82a876ecc11b5ca0977d1733adbe58599088a
runc version: 2d41c047c83e09a6d61d464906feb2a2f3c52aa4
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.12.4-1-ARCH
Operating System: Arch Linux
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 7.686GiB
Name: peregrin
ID: WGK4:K2IQ:GTX5:JMQQ:45EB:T3MU:OA62:TIUD:VIMX:VV3P:OISY:IRC6
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Code**
We tried to track it down in the code and found this section:
https://github.com/moby/moby/blob/361a473121f2cd4e688a3bb71fbeff00aaf674d7/daemon/graphdriver/btrfs/btrfs.go#L367-L368
It looks like the btrfs graph driver tries to enable/disable the quotas on a subvolume base. But this does not seem to be possible. https://github.com/torvalds/linux/blob/8c27cb3566762613a23c080e3db7d0501af9a787/fs/btrfs/qgroup.c#L946-L988

By the way, the procedure subvolEnableQuota() has the same issue.

### PR DESCRIPTION
**Pull request by vadorovsky**
Before this change, cleanup of the btrfs driver (occuring on each daemon
shutdown) resulted in disabling quotas. It was done with an assumption
that quotas can be enabled or disabled on a subvolume level, which is
not true - enabling or disabling quota is always done on a filesystem
level.

That was leading to disabling quota on btrfs filesystems on each daemon
shutdown.

This change fixes that behavior and removes misleading `subvol` prefix
from functions and methods which set up quota (on a filesystem level).

Fixes: #34593
Fixes: https://github.com/docker/for-linux/issues/78
Fixes: 401c8d176743 ("Add disk quota support for btrfs" (https://github.com/moby/moby/pull/19651) )
Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

**- What I did**

Ensured that moby deamon doesn't disable quotas in the btrfs filesystem.

**- How I did it**

- Removed the `subvolDisableQuota` method.
- Limited the `Cleanup` method just to unmounting.
- Removed the misleading `subbol` prefixes from methods which do operations on a filesystem level.

**- How to verify it**

- Ensure that btrfs quota is enabled (`btrfs quota enable /`)
- Start dockerd
- Gracefully shutdown dockerd
- Check if btrfs quota is still enabled by ensuring `btrfs qgroup show /` runs successfully.

I've done those steps inside the dev shell. They require btrfs-progs, which can be installed with `apt update && apt install btrfs-progs`.

**- Description for the changelog**

btrfs: Do not disable quota on cleanup

**- A picture of a cute animal (not mandatory but encouraged)**

doge

### CODE REVIEW

**Comment by cpuguy83** [comment]
It seems like the issue here may be that docker does not put itself into a subvolume.
And on systems such as opensuse, unless you actually give /var/lib/docker a subvolume, dockerd is on the subvolume for the rootfs.

I'm wondering if a better fix would be to give /var/lib/docker/btrfs its own subvolume.

**Comment by cpuguy83** [comment]
That said, I doubt we actually need to disable quota regardless.

**Review by kolyshkin — APPROVED** [review]
LGTM (disabling quota on unmount does not make sense).

File: daemon/graphdriver/btrfs/btrfs.go [hunk 1]
```diff
@@ -343,7 +335,7 @@ func (d *Driver) updateQuotaStatus() {
 	})
 }
 
-func (d *Driver) subvolEnableQuota() error {
+func (d *Driver) enableQuota() error {
```
Comment (tianon) [inline daemon/graphdriver/btrfs/btrfs.go:338]: (Somewhat orthogonally,) If the given path _isn't_ a subvolume, do you think we should have this function error out so we don't accidentally set a quota on `/` thinking we're setting it on something more specific?

**Comment by mrostecki** [comment]
@cpuguy83 @tianon After trying out your idea of enforcing `/var/lib/docker/btrfs` to be a subvolume, I got convinced that it's probably not a good one.

First of all, I don't think it's needed. Quotas are applied only on particular `/var/lib/docker/btrfs/subvolumes/[id]` subvolumes, here:

https://github.com/moby/moby/blob/62b5194f627571c019907edb11bb7bc2edd59933/daemon/graphdriver/btrfs/btrfs.go#L556-L558

https://github.com/moby/moby/blob/62b5194f627571c019907edb11bb7bc2edd59933/daemon/graphdriver/btrfs/btrfs.go#L668-L670

(these are the only places I found which call `subvolLimitQgroup()`)

I don't see any part of the driver which would apply quota on any "higher" directory, thus I don't see any danger of accidentally applying quota on the rootfs subvol even if there is no `/var/lib/docker/btrfs` subvolume and `subolumes/[id]` are children of rootfs.

Secondly, to enforce `/var/lib/docker/btrfs` as a subvolume, we would have to provide some safe way of migrating old users. There is no possibility in btrfs to convert an existing directory to a subvolume. The only possible solution would be creating a new subvolume, reflinking the old contents (reflink is the only way to "move" files into another subvolume without copying and wasting the space), then removing the old directory.

That migration would look like:

```
mv /var/lib/docker/btrfs /var/lib/docker/btrfs.bak
btrfs subvolume create /var/lib/docker/btrfs
cp -a --reflink=always /var/lib/docker/btrfs.bak/* /var/lib/docker/btrfs
rm -rf /var/lib/docker/btrfs.bak
```

The biggest problem, which finally convinced me to not do that, is the `cp --reflink` command. It takes a long time to execute it on large amount of files and implementing it in Go would be hard. What that command does, is in fact calling the FICLONE ioctl on every file.

There is a Go library which does that, but only on single files:

https://github.com/KarpelesLab/reflink

If we really wanted to use it, we would have to combine it with `filepath.Walk`. But again, that would be a long operation, taking approximately several minutes. Definitely too long to do that just for initializing the graphdriver.

Please let me know if you agree with my thoughts or if I'm missing something.

For now, I pushed my old version of the PR branch which only removes the "disable quota" part of code, to fix #34593.

**Review by tianon — APPROVED** [review]
LGTM :+1:

(Thanks for testing `/var/lib/docker/btrfs` as a subvol and for the detailed write-up about why we probably shouldn't! :sweat_smile::heart:)

**Review by cpuguy83 — APPROVED** [review]
LGTM

Thanks!

**Pull request merged**